src/Sylius/Bundle/ApiBundle/ApiPlatform/Factory/MergingExtractorResourceMetadataFactory.php

```diff
@@ -0,0 +1,168 @@
+
+        if (null !== $resource['attributes'] || [] !== $this->defaults['attributes']) {
+            $resource['attributes'] = (array) $resource['attributes'];
+            foreach ($this->defaults['attributes'] as $key => $value) {
+                if (!isset($resource['attributes'][$key])) {
+                    $resource['attributes'][$key] = $value;
+                }
+            }
+        }
+
+        return $this->update($parentResourceMetadata ?: new ResourceMetadata(), $resource);
+    }
+
+    /**
+     * Returns the metadata from the decorated factory if available or throws an exception.
+     *
+     * @throws ResourceClassNotFoundException
+     */
+    private function handleNotFound(?ResourceMetadata $parentPropertyMetadata, string $resourceClass): ResourceMetadata
+    {
+        if (null !== $parentPropertyMetadata) {
+            return $parentPropertyMetadata;
+        }
+
+        throw new ResourceClassNotFoundException(sprintf('Resource "%s" not found.', $resourceClass));
+    }
+
+    /**
+     * Creates a new instance of metadata if the property is not already set.
+     */
+    private function update(ResourceMetadata $resourceMetadata, array $metadata): ResourceMetadata
+    {
+        foreach (['shortName', 'description', 'iri', 'itemOperations', 'collectionOperations', 'subresourceOperations', 'graphql', 'attributes'] as $propertyName) {
+            $propertyValue = $this->resolveResourceMetadataPropertyValue($propertyName, $resourceMetadata, $metadata);
+            if (null !== $propertyValue) {
+                $resourceMetadata = $resourceMetadata->{'with' . ucfirst($propertyName)}($propertyValue);
+            }
+        }
+
+        return $resourceMetadata;
+    }
+
+    /** @return mixed */
+    private function resolveResourceMetadataPropertyValue(
+        string $propertyName,
+        ResourceMetadata $parentResourceMetadata,
+        array $childResourceMetadata
+    ) {
+        $parentPropertyValue = $parentResourceMetadata->{'get' . ucfirst($propertyName)}();
+
+        $childPropertyValue = $childResourceMetadata[$propertyName];
+        if (null === $childPropertyValue) {
+            return $parentPropertyValue;
+        }
+
+        if (null === $parentPropertyValue) {
+            return $childPropertyValue;
+        }
+
+        if (is_array($parentPropertyValue)) {
+            if (!is_array($childPropertyValue)) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Invalid child property value type for property "%s", expected array',
+                    $propertyName,
+                ));
+            }
+
+            return $this->mergeConfigs($parentPropertyValue, $childPropertyValue);
+        }
+
+        return $childPropertyValue;
+    }
+
+    private function mergeConfigs(...$configs): array
+    {
+        $resultingConfig = [];
+
+        foreach ($configs as $config) {
+            foreach ($config as $newKey => $newValue) {
```

Comment:
I really think that this is a bad idea, replacing the configuration over merging would be easier and more obvious.